### ISSUE TITLE
Windows ARM support

### ISSUE THREAD

**Issue by peterdk** [musescore/MuseScore#23709]
### Your idea

I have a new Dell XPS 13 with Snapdragon X processor and thus running Windows 11 ARM. I don't see support yet for ARM on Windows in Musescore, although Linux releases do already support it. So therefore I would like to request adding Windows ARM builds.

### Problem to be solved

Improved performance and energy efficiency on ARM laptops that run Windows 11.

### Checklist

- [X] This request follows the [guidelines](https://github.com/musescore/MuseScore/wiki/Reporting-bugs-and-issues) for reporting issues
- [X] I have verified that this feature request has not been logged before, by searching the [issue tracker](https://github.com/musescore/MuseScore/issues) for similar requests

**Comment by Interface007** [comment]
Does MuScore run well on Windows 11 ARM with SnapDragon? Are there issues other than performance?
This would be valuable information for users who plan to buy a new laptop and must work with MuseScore.

**Comment by hohwille** [comment]
> Does MuScore run well on Windows 11 ARM with SnapDragon? Are there issues other than performance? This would be valuable information for users who plan to buy a new laptop and must work with MuseScore.

Hi, I also have a WinOnArm Laptop with Snapdragon X (X1P-42-100).
MuseScore x86 version works with emulation and I do not see any functional change of behaviour (no bugs or crashes I do not observe on my x86 machine).
Surely it would be awesome to have a native ARM64 download option to avoid emulation.
From my PoV it is less a performance (x86 does not run significantly slow) issue but rather a sustainability issue since surely emulation wastes more energy, drains more battery, etc.

**Comment by cbjeukendrup** [comment]
There's good news: https://github.blog/changelog/2025-04-14-windows-arm64-hosted-runners-now-available-in-public-preview/
This should make it feasible to create arm64 builds for Windows. However, we can only do so after updating to the latest version of Qt, and there are still some problems blocking that.

**Comment by cbjeukendrup** [comment]
Update: after switching to Qt 6.9, there are still a few more hurdles to overcome: we need pre-built arm64 versions of `libsndfile` and `zlib`. The latter is easy, but `libsndfile` will be a lot of work, because it has so many dependencies. Some of those have not been updated for quite a while, so it isn't even certain that they _can_ be compiled for arm64 windows.

**Comment by sgsuriano** [comment]
> Update: after switching to Qt 6.9, there are still a few more hurdles to overcome: we need pre-built arm64 versions of `libsndfile` and `zlib`. The latter is easy, but `libsndfile` will be a lot of work, because it has so many dependencies. Some of those have not been updated for quite a while, so it isn't even certain that they _can_ be compiled for arm64 windows.

Hello @cbjeukendrup, as far as I understand, it shall be quite easy using vcpkg:

- Bootstrap vcpkg:
`git clone https://github.com/microsoft/vcpkg.git
cd vcpkg
.\bootstrap-vcpkg.bat`

- install:
`.\vcpkg install zlib:arm64-windows libsndfile:arm64-windows`

- Integrate vcpkg with CMake
`.\vcpkg integrate install`

- tell CMake to pull vcpkg-managed ARM64 libraries via the option:
`-DCMAKE_TOOLCHAIN_FILE="C:/path/to/vcpkg/scripts/buildsystems/vcpkg.cmake"`

I run successfully all vcpkg-related commands, but I can't tell if they are enough, since I have issues compiling Musescore 4.5.2 because of missing references to OpenGL (Windows 11 on ARM on a Surface Pro 11, with VS 2022 Community and QT 6.9.1)

**Comment by hmartinez82** [comment]
Is it possible to build MuseScore without OpenGL at all?
If not, the Windows SDK does provide Opengl32.lib for linking, but to run the app you must install the `OpenCL™, OpenGL®, and Vulkan® Compatibility Pack` app from MS Store

**Comment by hmartinez82** [comment]
@sgsuriano  I found some linking problems too, in the very end. Mainly because it was trying to use these prebuilt binaries for x64: https://github.com/musescore/musescore_prebuild_win_deps/tree/main/libx64

I did a local patch to the CMake project files to not use those files and use vcpkg for everything inlcuding zlib and libsndfile.

I was able to build the binary, now it's not getting past the splash screen due to failing to load a QML file, but at least it linked!

The error I'm  seeing in DbgView is:
```
02:11:25.441 | ERROR | main_thread     | GuiApp::setupNewContext | Failed to load main qml file, err: "qrc:/qt/qml/MuseScore/AppShell/platform/win/Main.qml:78 Type WindowContent unavailable\nqrc:/qt/qml/MuseScore/AppShell/WindowContent.qml:103 Type NotationPage unavailable\nqrc:/qt/qml/MuseScore/AppShell/NotationPage/NotationPage.qml:289 Type LayoutPanel unavailable\nqrc:/qt/qml/MuseScore/InstrumentsScene/LayoutPanel.qml:207 Invalid property assignment: \"__currentRowItem\" is a read-only property\n" 
```

**Comment by hmartinez82** [comment]
Well. I got it working. I locally added a workaround for https://qt-project.atlassian.net/browse/QTBUG-144701 by adding `data: ` to this line:  https://github.com/musescore/MuseScore/blob/69af3e1f8f82b263e64ae4bd252147b219016cf5/src/instrumentsscene/qml/MuseScore/InstrumentsScene/LayoutPanel.qml#L207

<img width="2750" height="1666" alt="Image" src="https://github.com/user-attachments/assets/60d52cf7-3877-46a1-95ba-6911d08bcc7b" />

Native ARM64 build!!

**Comment by hmartinez82** [comment]
@sgsuriano  I'm going to open a PR to replace those two dependencies with vcpkg. I really disliked pre-built binaries with all my being! 😅

**Comment by cbjeukendrup** [comment]
FWIW, last time I checked, (at that point fellow-) maintainers were opposed to the usage of vcpkg because of its relative complexity for new contributors to set up, and because of a tendency to prefer in-house-built infrastructure, even if that means reinventing wheels. However it's worth trying again. But there is a chance that we'll need to think of a different approach. Pre-built binaries that were built on GitHub Actions with a reproducible workflow would already be much better than the pre-built binaries from ?? years ago of which nobody knows the origin anymore... 🙂

**Comment by hmartinez82** [comment]
@cbjeukendrup oh ok. I had no idea :(

@Tantacrul @igorkorsukov  who can add the ARM64 prebuilt binaries to https://github.com/musescore/musescore_prebuild_win_deps ? And are those all the ones required? It seems that the zip file in the S3 bucket used by  the Windows CI workflow has even more binaries.

**Comment by igorkorsukov** [comment]
@hmartinez82 I'm concerned that we probably won't be able to release an official build of Windows ARM. We don't have the equipment for testing, and adding another platform for testing is too expensive for us.

And enthusiasts don't need prebuilds; they can create them themselves or get them from somewhere, like you did. 
Zlib and SSL (for Google CrashPad) are currently used as pre-build dependencies for Windows.
libsndfile shouldn't be needed now, I need to make sure... it's not used now.

But I can also tell you that we're in the process of improving our dependency management infrastructure. We plan to have modes for each dependency:
* Source Code
* Prebuild
* System

This will take some time, but we're already in the process.

**Comment by cbjeukendrup** [comment]
Would it be an idea to release Windows ARM binaries under the same policy as the Portable builds? I.e. they are not tested in-house, but issues can be reported and fixed. 
An easy way to test Windows ARM binaries in a representative environment, is under Parallels Desktop on Apple Silicon Macs.

**Comment by igorkorsukov** [comment]
We'll likely either add pre-builds for Windows ARM in our new dependency management infrastructure, or we'll be able to simply enable "dependencies as source code" mode and compile everything together.

As for testing, I'm mainly concerned about Qt. I don't think, for example, that engraving will behave differently, but any interaction with the OS is concerning (including rendering). It's unclear how much effort we'll be able to devote to this. But if we even one release, it will need to be supported always; it will be difficult to skip or abandon.

Also, we use the same Qt version for all platforms, meaning we won't be able to quickly resolve issues by updating Qt. For example, right now we can't update to Qt 6.11 (we can't even update to Qt 6.10.3).

**Comment by cbjeukendrup** [comment]
Qt is not a concern anymore; Windows ARM support was moved out of Technology Preview, into fully supported, at Qt 6.8.

It seems likely that the demand for an ARM build of MuseScore will increase rather than decrease, given the growing availability of ARM devices.

Another argument is that we already offer arm64 builds for Ubuntu, so why would we not for Windows.

By the way, are we actually still using libsndfile at all? It seems like not, since https://github.com/musescore/MuseScore/pull/29755, but I might be mistaken.

**Comment by igorkorsukov** [comment]
> By the way, are we actually still using libsndfile at all? It seems like not, since https://github.com/musescore/MuseScore/pull/29755, but I might be mistaken.

Yes, it shouldn't be used.
In the next few days, I'll remove references to libsndfile from the build scripts.

**Comment by Master-Hash** [comment]
> Well. I got it working. I locally added a workaround for https://qt-project.atlassian.net/browse/QTBUG-144701 by adding `data: ` to this line:  https://github.com/musescore/MuseScore/blob/69af3e1f8f82b263e64ae4bd252147b219016cf5/src/instrumentsscene/qml/MuseScore/InstrumentsScene/LayoutPanel.qml#L207
> 
> <img width="2750" height="1666" alt="Image" src="https://github.com/user-attachments/assets/60d52cf7-3877-46a1-95ba-6911d08bcc7b" />
> 
> Native ARM64 build!!

Could you share your build's installer or portable zip here?

**Comment by hmartinez82** [comment]
https://1drv.ms/u/c/9e3dcd8000d2e3bb/IQD2SAF23bCIRqsxlOHmPJZnActLRLSwSwlJPQrW7MnWJSg?e=qUQWOk . It's just a zipped output of the CMake install 

I'll try to add reproducible builds in my fork.

**Comment by prayaas-a** [comment]
It's exciting to see progress here. Linking to this semi-relevant [discussion](https://github.com/orgs/musescore/discussions/33145) about whether Arm64EC might make more sense.

**Comment by robdnz1** [comment]
> https://1drv.ms/u/c/9e3dcd8000d2e3bb/IQD2SAF23bCIRqsxlOHmPJZnActLRLSwSwlJPQrW7MnWJSg?e=qUQWOk . It's just a zipped output of the CMake install
> 
> I'll try to add reproducible builds in my fork.

Thank you so much! Did you manage to get MuseScore running natively on ARM without emulation? I’d like to try it myself as well. Do I just need to download this folder and run the executable, or is there an updated version available somewhere?

Thanks again for your help!

**Comment by hmartinez82** [comment]
Yes, just download and run the main executable  in the bin folder. Yes, native ARM64. I mean it opens, and loaded one score. I don't know the app completely to test more things

**Comment by cbjeukendrup** [comment]
By now, the dependencies problem is mostly solved because dependencies are now managed by a custom system: https://github.com/musescore/muse_deps

It looks like it provides Windows ARM builds too, so it should now be near-trivial to create official ARM builds of MuseScore.

The question though is how useful those will be, because most VSTs and MuseHub/MuseSounds probably don't support ARM yet.

**Comment by Master-Hash** [comment]
> The question though is how useful those will be, because most VSTs and MuseHub/MuseSounds probably don't support ARM yet. 

1. something better than nothing. I'm using the third party build above, with no musehub or vst, and it's perfect for basic usage, which is already great.
2. use arm64ec for vst compatibility

**Comment by hmartinez82** [comment]
@Master-Hash Yes, I was going to say ARM64EC too, but the issue with ARM64EC is that it's not something that you just "use", you have to build **everything** (read, all the dependencies) as ARM64EC, and yes, for a big project that use plugins extensively like Muse, building the binaries as ARM64EC would be best thing.

**Comment by cbjeukendrup** [comment]
A problem with ARM64EC is that Qt [does not support it](https://doc.qt.io/qt-6/supported-platforms.html#windows). It looks like Qt 6.12 [will support it](https://doc-snapshots.qt.io/qt6-6.12/supported-platforms.html#windows) apart from some modules MuseScore doesn't use, albeit without providing binaries, but I don't expect Muse is going to use Qt 6.12 until MuseScore 5.1 or 5.2.

Anyway, anyone is welcome to create a PR to add arm64 builds to Windows CI right now; but I'm afraid that's unlikely going to be me, because I don't work at Muse anymore.

**Comment by prayaas-a** [comment]
As a user, my vote would be towards holding out for Arm64EC support because currently all plugins just work (emulated). The native Arm version being less capable could be confusing/off-putting for users. With Arm64EC, plugins that want to port will benefit from native performance/efficiency while still keeping the rest compatible.

**Comment by Jojo-Schmitz** [comment]
We could use ARM64 now and switch to (or add) ARM64EC once available?

**Comment by prayaas-a** [comment]
> We could use ARM64 now and switch to (or add) ARM64EC once available?

That's definitely a possibility, but it does create some logistical gotchas:
- Any plug-ins that port to Arm64 now won't load in the Arm64EC version of Musescore and will need to recompile again. Arm64EC can only load x64 or Arm64EC code, not pure Arm64. This can be mitigated by encouraging plug-ins to build Arm64X/chameleon-binaries that contain both Arm64 and Arm64EC versions. 
- How would users switch from one to the other? Ideally, at some point Arm64 users should just be migrated to the new native version via an update.

Ultimately, imo, it depends on whether there are closed source plug-ins and how important compatibility is to the project. If compatibility with closed source plug-ins (that may not be quick to port) is important, Arm64EC makes sense, otherwise Arm64.
